File: src/Payload/SystemRequest.php
<?php
declare(strict_types=1);

namespace SMSolver\Payload;


//use SMSolver\Core\ArrayOf;
use RuntimeException;
use JsonSerializable;
use SMSolver\Core\Axis;
use SMSolver\Core\Math\VectorUtils;
use SMSolver\Core\Models\Beam;
use SMSolver\Core\Models\Force;
use SMSolver\Core\Models\ForceType;
use SMSolver\Core\Models\Node;
use SMSolver\Core\Models\NodeType;
use SMSolver\Core\Models\Reaction;
use SMSolver\Core\Models\ReactionType;
use SMSolver\Utils\OutputInfo;

class SystemRequest implements JsonSerializable
{
    /** @var Node[] $nodes */
    private array $nodes = [];

    /** @var Beam[] $beams */
    private array $beams = [];

    /** @var Force[] $forces */
    private array $forces = [];

    /** @var Reaction[] $reactions */
    private array $reactions = [];

    private array $referenceSymbolMatrix = [];

    /**
     * @param array<string,array<array<string,scalar>>> $data
     * @return self
     */
    public static function constructFromArray(array $data): self
    {
        $instance = new self();

        foreach ($data['nodes'] as $nodeData) {
            $node = Node::constructFromArray($nodeData);
            $instance->addNode($node);
            unset($node);
        }

        foreach ($data['beams'] as $beamData) {

            $startNode = $instance->getNodeById((string)$beamData['startNode']);
            $beamData['_startNode'] = $beamData['startNode'];
            $beamData['startNode'] = $startNode;

            $endNode = $instance->getNodeById((string)$beamData['endNode']);
            $beamData['_endNode'] = $beamData['endNode'];
            $beamData['endNode'] = $endNode;

            $beam = Beam::constructFromArray($beamData);

            $startNode->addBeam($beam);
            $endNode->addBeam($beam);

            $instance->addBeam($beam);
            unset($beam);
        }

        foreach ($data['forces'] as $forceData) {

            $forceNode = $instance->getNodeById((string)$forceData['node']);
            $forceData['_node'] = $forceData['node'];
            $forceData['node'] = $forceNode;

            $force = Force::constructFromArray($forceData);

            $forceNode->addForce($force);

            $instance->addForce($force);
            unset($force);
        }

        return $instance;
    }

    public function jsonSerialize()
    {
        $arrayRepresentation = [];
        $arrayRepresentation['nodes'] = $this->nodes;
        $arrayRepresentation['beams'] = $this->beams;
        $arrayRepresentation['forces'] = $this->forces;
        return $arrayRepresentation;
    }

    // Getters

    public function getNodeById(string $nodeId): Node
    {
        return $this->nodes[$nodeId] ?? throw new RuntimeException('Node not found by id: ' . $nodeId);
    }

    public function getBeamById(string $beamId): Beam
    {
        return $this->beams[$beamId] ?? throw new RuntimeException('Beam not found by id: ' . $beamId);
    }

    public function getForceById(string $forceId): Force
    {
        return $this->forces[$forceId] ?? throw new RuntimeException('Force not found by id: ' . $forceId);
    }


    // Adders

    public function addNode(Node $node): void
    {
        $this->nodes[$node->getId()] = $node;
    }

    public function addBeam(Beam $beam): void
    {
        $this->beams[$beam->getId()] = $beam;
    }

    public function addForce(Force $force): void
    {
        $this->forces[$force->getId()] = $force;
    }

    /**
     * @return float[][]
     */
    public function generateMatrix(): array
    {
        $n = $this->calcN();

        $this->buildReferenceSymbolMatrix();

//        OutputInfo::printJSONln($this->referenceSymbolMatrix);
//        die();
        $c = count($this->referenceSymbolMatrix);

        $Ax = VectorUtils::zerosMatrix($c - 1, $c);

        $i = 0;
        foreach ($this->nodes as $node) {

//            OutputInfo::echoln('using node ' . $node->getId() . ' with $i: '.$i);
            //X for $node
            $valuesBySymbol = $node->getValuesBySymbolByAxis(Axis::X());
            foreach ($valuesBySymbol as $symbol => $value) {
                $Ax[$i][$this->referenceSymbolMatrix[$symbol]] = $value;
            }

            //Y for $node
            $valuesBySymbol = $node->getValuesBySymbolByAxis(Axis::Y());
            foreach ($valuesBySymbol as $symbol => $value) {
                $Ax[$i + 1][$this->referenceSymbolMatrix[$symbol]] = $value;
            }
            $i += 2;
        }
//        echo json_encode($Ax);
//        die();
//
//        var_dump($valuesBySymbol);
//        die();
//        var_dump($this->referenceSymbolMatrix);

        return $Ax;
    }

    private function buildReferenceSymbolMatrix(): void
    {

        // build $reactions array from nodes, beams and incognito force
        // then build referenceSymbolMatrix from this array

        if (empty($this->reactions)) {
            foreach ($this->beams as $beam)
                $this->reactions[] = Reaction::constructFromBeam($beam);

            foreach ($this->nodes as $node)
                foreach (Reaction::constructFromNode($node) as $reaction)
                    $this->reactions[] = $reaction;

            foreach ($this->forces as $force)
                if (ForceType::UNKNOWN()->equals($force->getType()))
                    $this->reactions[] = Reaction::constructFromForce($force);

            $this->reactions[] = Reaction::resultReaction();
        }

        foreach ($this->reactions as $intIndex => $reaction)
            $this->referenceSymbolMatrix[$reaction->getSymbol()] = $intIndex;
    }

//    private function buildBaseMatrix(): array
//    {
//
//    }

    private function calcN(): int
    {
        $n = 0;

        foreach ($this->nodes as $node)
            $n += $node->getN();

        return $n;
    }

    public function getReferenceSymbolMatrix(): array
    {
        return $this->referenceSymbolMatrix;
    }

    /**
     * @param array $result
     * @return Reaction[]
     */
    public function mapReactionsWithResults(array $result): array
    {
        $reactions = array_filter($this->reactions,
            fn(Reaction $reaction) => !ReactionType::RESULT()->equals($reaction->getType())
        );
        foreach ($reactions as $intIndex => $reaction)
            $reaction->setMagnitude($result[$intIndex]);

        return $reactions;
    }
}

File: src/Payload/SystemRequest.go
package Payload

import (
	. "static_mechanism_solver/src/Core"
	. "static_mechanism_solver/src/Core/Math"
	. "static_mechanism_solver/src/Core/Models"
)

type SystemRequest struct {
	nodes                 map[string]*Node
	beams                 map[string]*Beam
	forces                map[string]*Force
	reactions             []Reaction
	referenceSymbolMatrix map[string]int
}

func ConstructSystemRequestFromArray(data SystemData) SystemRequest {
	s := SystemRequest{}
	s.referenceSymbolMatrix = map[string]int{}
	for nodeId, nodeData := range data.Nodes {
		n := ConstructNode(NodeType(nodeData.Type), nodeId, nodeData.X, nodeData.Y)

		s.AddNode(&n)
	}

	for beamId, beamData := range data.Beams {
		startNode := s.getNodeById(beamData.StartNode)
		endNode := s.getNodeById(beamData.EndNode)
		b := ConstructBeam(*startNode, *endNode, beamId)

		startNode.AddBeam(&b)
		endNode.AddBeam(&b)
		s.AddBeam(&b)
	}

	for forceId, forceData := range data.Forces {
		forceNode := s.getNodeById(forceData.Node)
		f := ConstructForce(ForceType(forceData.Type), forceId, forceData.Magnitude, forceData.Angle)

		forceNode.AddForce(&f)
		s.AddForce(&f)
	}

	return s
}

func (s SystemRequest) getNodeById(nodeId string) *Node {
	x, ok := s.nodes[nodeId]
	if !ok {
		panic("Node not found by id: " + nodeId)
	}
	return x
}

func (s SystemRequest) getBeamById(beamId string) *Beam {
	x, ok := s.beams[beamId]
	if !ok {
		panic("Beam not found by id: " + beamId)
	}
	return x
}

func (s SystemRequest) getForceById(forceId string) *Force {
	x, ok := s.forces[forceId]
	if !ok {
		panic("Force not found by id: " + forceId)
	}
	return x
}

func (s *SystemRequest) AddNode(n *Node) { // _todo: maybe replace with n.Beams = append(n.Beams, b)
	if s.nodes == nil {
		s.nodes = map[string]*Node{}
	}
	s.nodes[n.Id] = n
}

func (s *SystemRequest) AddBeam(b *Beam) { // _todo: maybe replace with n.Beams = append(n.Beams, b)
	if s.beams == nil {
		s.beams = map[string]*Beam{}
	}
	s.beams[b.Id] = b
}
func (s *SystemRequest) AddForce(f *Force) { // _todo: maybe replace with n.Beams = append(n.Beams, b)
	if s.forces == nil {
		s.forces = map[string]*Force{}
	}
	s.forces[f.Id] = f
}

// _todo: remove this
func (s *SystemRequest) addReaction(r Reaction) {
	s.reactions = append(s.reactions, r)
}

// _todo: remove this
func (s SystemRequest) PrintReactions() string {
	text := " "
	for _, r := range s.reactions {
		text += r.Print()
	}
	return text
}

func (s *SystemRequest) GenerateMatrix() [][]float64 {
	//n := s.calcN()
	s.buildReferenceSymbolMatrix()
	c := len(s.referenceSymbolMatrix)
	Ax := ZerosMatrix[float64](c-1, c)
	i := 0
	k := 0
	for _, node := range s.nodes {
		// X for node
		valuesBySymbol := node.GetValuesBySymbolByAxis(X)
		for symbol, value := range valuesBySymbol {
			Ax[i][s.referenceSymbolMatrix[symbol]] = value
		}
		k += 1
		// Y for node
		valuesBySymbol = node.GetValuesBySymbolByAxis(Y)
		for symbol, value := range valuesBySymbol {
			Ax[i+1][s.referenceSymbolMatrix[symbol]] = value
		}
		i += 2
	}
	return Ax
}
func (s *SystemRequest) buildReferenceSymbolMatrix() {
	// build $reactions array from nodes, beams and incognito force
	// then build referenceSymbolMatrix from this array
	if len(s.reactions) == 0 {
		for _, b := range s.beams {
			s.reactions = append(s.reactions, ConstructFromBeam(*b))
		}
		for _, n := range s.nodes {
			for _, r := range ConstructFromNode(*n) {
				s.reactions = append(s.reactions, r)
			}
		}
		for _, f := range s.forces {
			if f.ForceType == Unknown {
				s.reactions = append(s.reactions, ConstructFromForce(*f))
			}
		}
		s.reactions = append(s.reactions, ResultReaction())
	}
	for i, r := range s.reactions {
		s.referenceSymbolMatrix[r.GetSymbol()] = i
	}
}
func (s SystemRequest) calcN() int {
	n := 0
	for _, node := range s.nodes {
		n += node.GetN()
	}
	return n
}

// func (s SystemRequest) getReferenceSymbolMatrix
func (s SystemRequest) MapReactionsWithResults(result []float64) []Reaction {
	var rs []Reaction
	for _, r := range s.reactions {
		if r.GetType() != R_Result {
			rs = append(rs, r)
		}
	}
	for i := range rs {
		rs[i].SetMagnitude(result[i])
	}

	return rs
}
